### PR TITLE
Cleanup temporary files better

### PR DESCRIPTION
**Pull request by brianphaley**
Changed hotsos to setup its global temp directory
before setting up logging so that all files can
live in one place. Removal is now done in the code in cli() as well.

Code in client.py was cleaned-up accordingly as it was no longer needed.

Resolves: #730

### CODE REVIEW

**Review by dosaboy — COMMENTED** [review]
I guess this calls into question what the purpose of the temporary log file is. With this change the logfile gets deleted when hotsos finished and the environment is torn down so does not give the possibility of post-analysis. Having said that, leaving the file after each run is not ideal. Also in this change the teardown is not within a try/except so if the main plugin runner fails the teardown will not happen. I'd like to propose the following:
  * make sure teardown_env is within a try/except like it was originally so that we always delete the plugin data (which can get quite large)
  * leave the logfile where it is BUT delete it at the same time as teardown_env() IFF there is no exception raised otherwise print a message saying e.g. "an error occurred while running hotsos see logfile at XXX for details"

File: hotsos/client.py [hunk 1]
```diff
@@ -410,4 +403,3 @@ def run(self):
                         self.summary.update(plugin, content.get(plugin))
         finally:
             log.name = 'hotsos.client'
-            self.teardown_global_env()
```
Comment (brianphaley) [inline hotsos/client.py:413]: So I just wanted to make it clear this call *mostly* removed the temp files, but for some reason these kernel plugin ones were always left behind:

$ find tmpzssaz5uq
tmpzssaz5uq
tmpzssaz5uq/kernelva2q73j_
tmpzssaz5uq/kernelva2q73j_/locks
tmpzssaz5uq/kernelva2q73j_/locks/cache_commands.lock
tmpzssaz5uq/kernelva2q73j_/locks/cache_all_global.lock
tmpzssaz5uq/kernelva2q73j_/caches
tmpzssaz5uq/kernelva2q73j_/caches/host_helpers_cli
tmpzssaz5uq/kernelva2q73j_/caches/host_helpers_cli/commands
tmpzssaz5uq/kernelva2q73j_/caches/host_helpers_cli/commands/hostname

And 'kernel' is the last one run based on PLUGIN_RUN_ORDER.

Which actually led me to change the order and see what happens... and it shows the last one run always leaves these same files behind.

So maybe there is some better way to fix this?

I actually found the real problem. Above in get() there is this call:
    CLIHelper().hostname()

Which is going to write to the hostname file in the temp dir. Since HotSOSConfig.plugin_tmp_dir is set it writes into the last plugins temp dir, which in all cases is the kernel one, re-creating some of the tree there. Clearing that variable solves the issue, at least for the directory. That was not obvious at all.

I'll look at the log file comment you left and see how to deal with that.

File: hotsos/client.py [hunk 2]
```diff
@@ -380,6 +387,19 @@ def setup_plugin_env(self, plugin):
         HotSOSConfig.plugin_tmp_dir = tempfile.mkdtemp(prefix=plugin,
                                                        dir=global_tmp)
 
+    def clear_plugin_env(self):
+        log.debug("clearing plugin env")
+        HotSOSConfig.plugin_tmp_dir = None
```
Comment (dosaboy) [inline hotsos/client.py:392]: this should go into teardown_global_env()

File: hotsos/client.py [hunk 3]
```diff
@@ -408,6 +429,15 @@ def run(self):
...
+            if temp_log:
+                print('See temp log file for possible details:', temp_log)
+            raise
         finally:
             log.name = 'hotsos.client'
+            self.clear_plugin_env()
             self.teardown_global_env()
+
+        # This will only get called if no exceptions
+        self.remove_temp_log_file()
```
Comment (dosaboy) [inline hotsos/client.py:443]: this can go inside the ```try:``` clause at the end since we are executing it if no exception occured

File: hotsos/client.py [hunk 4]
```diff
@@ -371,7 +378,7 @@ def setup_global_env(self):
     def teardown_global_env(self):
         log.debug("tearing down global env")
         if os.path.exists(HotSOSConfig.global_tmp_dir):
-            shutil.rmtree(HotSOSConfig.global_tmp_dir)
+            shutil.rmtree(HotSOSConfig.global_tmp_dir, ignore_errors=True)
```
Comment (dosaboy) [inline hotsos/client.py:381]: why would we need ignore_errors?

File: hotsos/client.py (hunk 2 above)
Comment (brianphaley) [reply hotsos/client.py:392]: Done

File: hotsos/client.py (hunk 3 above)
Comment (brianphaley) [reply hotsos/client.py:443]: Done

File: hotsos/client.py [hunk 5]
```diff
@@ -284,6 +284,13 @@ def get(self, fmt='yaml', html_escape=False, minimal_mode=None,
 
         return filtered
 
+    def temp_log_filename(self):
```
Comment (dosaboy) [inline hotsos/client.py:287]: nit: i feel like this should live in hotsos.core.log but can do as a separate patch

**Review by dosaboy — APPROVED** [review]
